### ISSUE TITLE
ERR_REQUIRE_ESM error

### ISSUE THREAD

**Issue by joelmoss** [benmosher/eslint-plugin-import#2050]
My webpack config is an esmodule, so because eslint-plugin-import-webpack is using `require()`, I get:

`Resolve error: Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: ./webpack.config.js
require() of ES modules is not supported.`

And things fail.

Any way you can support ESM webpack config files?

**Comment by ljharb** [comment]
I'm not sure how we possibly could (nor why it'd be a good idea for you to have a config file be ESM). We support versions of node that don't have `import()`, but also, eslint requires everything be synchronous, and ESM can only be asynchronously consumed.

**Comment by joelmoss** [comment]
Well webpack 5 supports ESM for its config files, and the latest Node LTS now supports and recommends ESM, so I don't think its unreasonable to support that.

> eslint requires everything be synchronous, and ESM can only be asynchronously consumed.

Think I may have confused you with my mention of `import()`, which is the async version of `import`. I actually mean regular sync `import`s.

**Comment by ljharb** [comment]
That can only be used in ESM itself, so it would be a major rewrite of this plugin to be able to do that, for virtually no benefit except allowing ESM webpack config files, and with major downsides, including a semver-major that dropped the majority of our supported eslint and node versions.

**Comment by ljharb** [comment]
Since eslint itself doesn't support plugins or rules being written in native ESM, it's not something that's possible now.

I very much doubt the node community will "move" to native ESM wholesale - using it has lots of downsides, and CJS isn't "legacy" and is never going away completely.

**Comment by joelmoss** [comment]
Perhaps you're right, but I reckon ESM may become the dominant standard.

Just my two cents worth 😏

**Comment by ljharb** [comment]
As one of the members of the node modules working group, I am very skeptical that will ever come to pass.

Either way, until eslint supports plugins/configs being in native ESM - or until they support them being asynchronous - there's nothing anyone in the eslint community can do here, so I'll close this for now.

**Comment by kemaakpala** [comment]
@ljharb is there anyway to get arround this, apart from having to revert my webpack config files.

**Comment by ljharb** [comment]
@kemaakpala no. Stick to CJS (or transpiled ESM) for many reasons, including this issue.

**Comment by guillaumebrunerie** [comment]
I'm also having this issue, I would like to have my Webpack config file written using ES modules as I see no good reason for having to learn and use two module systems in the same project.

I managed to fix it by removing `eslint-import-resolver-webpack` and replacing it with `eslint-import-resolver-typescript` which does not seem to have this problem.

**Comment by JanJakes** [comment]
@ljharb Do you think it would be a reasonable idea to allow for passing the Webpack config content directly? That might allow users to transpile the config in any way they want. E.g., something like (pseudocode):

```js
settings: {
  'import/resolver': {
    webpack: {
      config: babel.transpile(myWebpackConfig),
    },
},
```

I feel like that could solve the issues for those who are willing to make a bit of additional effort. The issue is that once you have a `"type": "module"` package, it's not a great experience to mix the two module types. In that case, for instance, Webpack will expect your config to be using ESM.

**Comment by JanJakes** [comment]
Ahh, nevermind. I found out passing a config object is already possible, so something like the following just works:

```ts
import webpackConfig from './webpack.config.js';

...

settings: {
  'import/resolver': {
    webpack: { config: webpackConfig },
  },
},
```

Yep, it needs to be a static (sync) import, but it's a decent workaround.
